Feature/issue 052

102092 opened the pull request:
Resolved #52 

- Post FindById 기능 추가
- 해당 기능 APi Docs 추가
- Post 생성, 응답 DTO 객체 생성, 수정

102092 committed:
[#52] Feat : api-guide.adoc findById API 추가

102092 committed:
[#52] Refactor : import optimize

- 쓸 데 없는 의존성 삭제

102092 committed:
[#52] Feat : findById 테스트 추가

- 포스트 메서드 이름 수정, spring rest docs 이름에서 삭제
- findByID 테스트 추가

102092 committed:
[#52] Refactor : Post Create Request, Response 리팩토링

- 클래스명 수정. 좀 더 명료하도록
- CreateResponse 응답 객체 생성
- Post Entity 내부 메서드 수정
- controller 메서드명 수정 controller만 보고도, 내부 동작을 알 수 있도록

hanurii commented on src/main/java/com/codesquad/rare/domain/post/Post.java:
```diff
@@ -40,10 +41,10 @@
   private String title;
 
-  //TODO content nullable, length 설정
+  @Lob
```
@Lob 은 또 뭘까요 정리해볼게 늘어났습니다!

102092 replied:
대용량 데이터를 담게 해주는 어노테이션으로 알고있습니다만.. 저도 정리가 필요할듯 하네요

hanurii replied:
https://www.notion.so/han95210/Annotation-981c64bdcf4349028baa9a10905c5a46#8213b036cadd4321a8037b1c77b4cb28

ㅋㅋ 궁금해서 정리해봤습니다

hanurii commented on src/main/java/com/codesquad/rare/domain/post/Post.java:
```diff
@@ -61,15 +62,15 @@
 
   private LocalDateTime createdAt;
 
-  public static Post from(PostRequestDto postRequestDto) {
+  public static Post toEntity(PostCreateRequest postCreateRequest, Account author) {
```
조졸두님이 쓰셨다니 할말이 없었습니다. 그치만 저 틀린건 아니죠? ㅋㅋㅋㅋ...??

102092 replied:
넹넹 저도 Dto라는 이름보다는 좀 더 명료한 클래스명이 좋아서 저렇게 사용하고 있어요!
하밀은 어떤 네이밍이 괜찮으신가요?

hanurii replied:
아 Dto 네이밍이 아니라 정적 팩터리 메서드 패턴 네이밍 컨벤션에 관한걸 여쭤보는 코멘트였습니다.
from -> toEntity 로 변경하셨길래요 ㅋㅋ

102092 replied:
ㅋㅋ 써니에게서 배웠습니당!!

hanurii commented on src/main/java/com/codesquad/rare/domain/post/request/PostCreateRequest.java:
```diff
@@ -1,28 +1,26 @@
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
 @ToString
-public class PostRequestDto {
+public class PostCreateRequest {
```
평소에 Dto라는 표현을 잘 안쓰시는 편인가요? 의견이 궁금합니다

hanurii replied:
클래스 네이밍 컨벤션에도 동사형이 들어가나요? 궁금합니다!

102092 replied:
네이밍 컨벤션에 동사형이 들어가도 되는지는 잘 모르겠어요
그렇지만, 저는 객체의 역할을 확실히 나타내주기 위해서 Create, Reqeust, Response, Modify등등으로 명시해주는 편이에요 
Dto란건 좀 너무 역할이 모호하다고 생각해서요..

hanurii replied:
ㅋㅋ 저도 네이밍 컨벤션이 정확히 어떻게 사용되는지 궁금해서 찾아봤습니다
https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

클래스 이름은 일반적으로 명사 또는 명사구로 표현해야하고 최대한 약어와 두문자어(HTML, URL과 같이 통상적으로 쓰이는 두문자어는 제외)를 사용하지 않고 전체 단어를 다 쓰라고 그러네요 ㅋㅋㅋ 그치만 저는 영어를 잘 몰라서 명사와 명사구로 이름 짓는 걸 못하니... 이번 프로젝트는 이대로 가시죠? ㅋㅋㅋㅋ

hanurii replied:
DTO 네이밍은 항상 어려운 것 같습니다. 저도 DTO란 접미사를 붙여야하나... 고민도 됩니다. 

지난 번 단이 올려준 외국에서 DTO 네이밍을 사용하는 형태를 올려줬는데 이런 형태의 사용법도 있다네요. (저도 아직 사용안해봤습니다 ㅋㅋ)  https://cassiomolin.com/2016/02/11/give-better-names-to-your-dtos/ 심심하실 때 읽어보세요

hanurii commented on src/main/java/com/codesquad/rare/domain/post/PostController.java:
```diff
@@ -24,27 +25,23 @@
   @GetMapping
-  public ApiResult<List<Post>> main() {
-    return OK(postService.findAll());
+  public ApiResult<List<Post>> findAllOrderByCreateAt() {
+    return OK(postService.findAllAndOrderByCreatedAtDesc());
```
모든걸 불러오는데 생성된 순서대로 불러오라 이런 뜻인가요? 영어 짱 어렵다...

최신순으로 정렬되는 걸 디폴트로 저희끼리 생각해놓고 사용하는 건 어떠신가요? (그럼으로써 메서드명 간략화를 기대) 

원래 디폴트로 정렬되는 기준이 최신순이 아니죠? 생성순이죠?

102092 replied:
음 저도 보니까, 컨트롤러의 메서드가 너무 긴 것 같네요.
저는 포스트를 다 불러오는데, 생성시간 오름차순으로, 그러니까 최신순으로 정렬되어서 불러오는 것을 의도한 건데..
어렵네요
간단하게 `latest` 로 바꿔볼까요?
어떻게 생각하시나요??

hanurii replied:
In Order 가 순서대로라는 뜻이라는데
In Latest Order == 최신 순서대로
이런 의미로 사용해보는 건 어떠세용?

102092 replied:
어 좋습니다 ㅎㅎ
`findAllPostInLatestOrder` 이 이름으로 할까요?

hanurii replied:
넵!

hanurii commented on src/main/java/com/codesquad/rare/domain/post/request/PostCreateRequest.java:
```diff
@@ -1,28 +1,26 @@
   private String thumbnail;
 
-  private Account author;
+  private Long authorId;
```
이거 김영한님 강의에서 얼핏 본거 같은데 전체 객체를 가져오지 말고
그 객체의 식별자값만 가져오라고 했던건가... 그거 맞나요??

102092 replied:
제가 Long Type으로 바꾼 이유는, 
Jwt 토큰 방식등으로 회원 인증이 아직 만들어지지 않은 상태에서,
포스트 생성 요청을 받을 때는, 어떤 회원이 요청하는 지 알아야 되어서 넣어놨습니다.
Controller에서 @RequestBody로 매칭되는 만큼, 제일 간소하게 DTO 객체를 만들고 싶었어요

Front 요청 --> Controller 에서 받아서 --> AuthorId를 통해 해당 Entity를 찾고 이를 기반으로 Post Entity를 만드는 순서로
진행되어야 한다고 생각했습니당

hanurii replied:
아하 제가 다른 거랑 조금 착각했습니다 이해했습니다!!

hanurii commented on src/main/java/com/codesquad/rare/domain/post/response/PostCreateResponse.java:
```diff
@@ -0,0 +1,16 @@
+@Setter
+@NoArgsConstructor
+@ToString
+public class PostCreateResponse {
```
클래스 네이밍 컨벤션에도 동사형이 들어가나요? 궁금합니다!

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -52,11 +59,20 @@
   MockMvc mockMvc;
 
   @MockBean
-  PostService postService;
+  PostController postController;
```
PostService 를 PostController 로 변경하면서 생기는 이슈는 없었나요? 궁금합니다. 없으면 없다고 해주셔도 됩니다.

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -52,11 +59,20 @@
+        .build();
+  }
+
   public static String asJsonString(final Object obj) {
```
파라미터 안에 final이 들어간 형태는 처음보네요 메서드가 끝날 때까지 파라미터 값이 변경되지 않는 걸로 이해하면 될까요?

hanurii replied:
찾아보니 로직상 변경되면 안되는 경우 final로 값을 고정시켜준다고 하네요 하나 배웠습니다!

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -157,98 +172,176 @@ void find_all_posts() throws Exception {
   @DisplayName("포스트 생성")
   @Test
-  public void create_post_spring_rest_docs() throws Exception {
+  void create_post() throws Exception {
```
테스트 코드 메서드 네이밍 컨벤션이 있더라구요. 한번 차용해서 사용해보면 어떨까 건의드립니다. 문서를 읽는 비용을 줄이는 하나의 작업이라고 생각해주세요 ㅋㅋ
http://blog.naver.com/PostView.nhn?blogId=genycho&logNo=220317775484&parentCategoryNo=&categoryNo=73&viewDate=&isShowPopularPosts=true&from=search

102092 replied:
우선 좋은 문서 공유 감사드립니다
저희는 @DisplayName 를 통해, 테스트 코드를 나타내는 만큼
한글로 표현해보면 어떨까요?

메서드 이름은, 테스트 하려는 가장 핵심적인 기능만을 작성하고요..

예를 들면,

```java
@DisplayName(" PostCreateRequest를 통하여 Post를 생성할 수 있다")
void create_post() ...
```
이런 식으로요!
혹시 생각하시는 방법이 있으시면 말씀해주세용

hanurii replied:
오오 ㅋㅋㅋ 좋아요 굿아이디어입니다 영어시렁...

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -157,98 +172,176 @@ void find_all_posts() throws Exception {
+    response.setPostId(1L);
+    log.debug("response : {}", response);
+    log.debug("create : {}", postCreateRequest);
+    given(postController.save(any())).willReturn(OK(response));
```
아까 간략히 들었지만 any() 의 쓰임새가 궁금하네요. mokito 에서만 사용되는 메서드인가요?
아까 설명을 들을 땐 '어떠한 것이 들어와도 괜찮다' 라고 이해했는데 맞나요?

102092 replied:
저도 정확히는 이해하지 못했습니다.. 너무 늦게 알게 된 지라..
우선 #53 에 작성된 내용을 봐주셨으면 하고,

제가 이해한 바로는 `any()` 아마도, 어떤 것들이 들어와도 `willReturn(..) `에 명시된 객체를 리턴하라는 의미라 생각했어요.

작성된 오류가 jsonString으로 변환된 것이 .save(여기) 들어가는 객체와 같다고 매칭이 되지 않아서
오류가 발생하더라구여..

hanurii replied:
any() 쓰임새를 떠나서 given 을 어떻게 사용하는지에 따라서 오류가 정말 많이 나는 것 같습니다. given에 관련된 트러블 슈팅도 정리해보고 싶네요 ㅎㅎ

102092 replied:
진행할수록 정리해야하는 것들이 느는 군요 ㅠㅠ

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -157,98 +172,176 @@ void find_all_posts() throws Exception {
-   }
+    result
+        .andExpect(status().isOk())
+        .andDo(print())
+        .andDo(
+            document("{class-name}/{method-name}",
+                getDocumentRequest(),
+                getDocumentResponse(),
```
andDo(print) 의 역할이 궁금하네요. prettyPrint() 와는 다른건가요??

102092 replied:
테스트 로그에 print해주는 역할로 알고 있어요
prettyPrint는 아마도? restdocs에 보기 좋게 뿌려주는 걸로 알고있습니다

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -157,98 +172,176 @@ void find_all_posts() throws Exception {
+                getDocumentRequest(),
+                getDocumentResponse(),
+                requestFields(
+                    fieldWithPath("title")
+                        .description("포스트 제목")
+                        .type(JsonFieldType.STRING),
```
가독성 차원에서 너무 코드가 길어지니까 한 눈에 보기 어려워서 밑에 방식을 우선 사용하고 있는데 위 방식을 사용하는 한님 생각은 어떠신지 궁금합니다!!

```
fieldWithPath("title").description("포스트 제목").type(JsonFieldType.STRING),
fieldWithPath("title").description("포스트 제목").type(JsonFieldType.STRING),
fieldWithPath("title").description("포스트 제목").type(JsonFieldType.STRING)
```

102092 replied:
좋습니다. 수정해서 올릴게요 ㅎㅎ

hanurii replied:
감사합니다!

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -157,98 +172,176 @@ void find_all_posts() throws Exception {
         .build();
 
-    Post post = Post.builder()
+    Post post1 = Post.builder()
```
post로 하셨다가 post1로 하신 이유가 있을까용?

102092 replied:
아마도 리팩토링 하는 과정에서, 저도 모르게 변경한듯 하네요.. 이부분은 수정해서 재 커밋할게요 ㅠㅠ

hanurii replied:
감사합니다~

hanurii reviewed (commented):
너무 고생하셨습니다!! 아까 많이 고군분투하시던데 잘 해결되어서 다행입니다 수요일 잘 쉬시고 목요일날 뵙겠습니다 ^^7
(저는 좋아요순 페이징 기능 만들어보고 있을게요 곧 PR 날리겠습니다)

102092 committed:
[#52] Refactor : code reformatting, 변수명 수정

- 가독성을 높히기 위해서, 코드 리포멧
- 사용 하지 않는 객체 삭제
- 이상한 변수명 수정

102092 committed:
[#52] Refactor : PostController 메서드 이름 수정

- 해당 메서드가 어떤 역할을 하고 있는지 정확하게 나타내기 위해서
- ...latest -> 최신순으로 모든 것을 찾아 주는 역할
- create -> Post를 생성하는 역할

hanurii commented:
고생하셨습니다 머지할게요
